Add Dev Blog comments via Discourse API

john-rock opened the pull request:
## Description & motivation
[Asana task](https://app.asana.com/0/1200099998847559/1203158602320485/f)

This PR is a continuation of a prior PR that was already in progress on adding comments via the Discourse iframe https://github.com/dbt-labs/docs.getdbt.com/pull/2218

🟢 [Component Documentation](https://www.notion.so/dbtlabs/Dev-Blog-Comments-Component-c2d7a183f57549f1989d099e48d3f4c2) in Notion 🟢

This PR will add comments to blog posts via the Discourse API.
Current behavior:
1. When a dev blog post is loaded check to see if the matching topic exists in Discourse (currently matching on the post title).
2. If it does not exist in Discourse, we take the name of the blog post and add a new topic in Discourse with the name of the blog post
3. Else we query the API to get the Discourse external_id that matches the name of the blog post and return the comments for that specific topic.

Data returned for each comment via API currently is listed below. There are a bunch of other properties we can retrieve from the API so please let me know if you would like to see more or less data returned.
- Username
- User Title
- Comment

## 🧪 Testing this feature
There are a few tests we can perform to make sure this is working as expected.

### 1. Comments are being loaded successfully
- Visit https://deploy-preview-2577--docs-getdbt-com.netlify.app/blog/how-to-create-near-real-time-models-with-just-dbt-sql#write-idempotent-models and confirm all test comments are loading.

### 2. A new post is generating the correct topic in Discourse.
- Visit any blog post. 
- Comments section should display "Start Discussion" CTA.
- Click "Start Discussion" and confirm you are linked out to the topic in Discourse
![image](https://user-images.githubusercontent.com/46692803/208489183-16b8d804-2b9f-474c-a6f3-886b500551a3.png)

### 3. Comments added in Discourse are rendered on the Dev Blog post
- Visit any blog post.
- Click either Start Discussion (if no comments exist yet) or Continue Dicussion (if comments do exist).
- Once the topic opens in Discourse, write a comment.
- Navigate back to Dev Blog post, refresh, and confirm the new comment is rendered.

[Netlify function log](https://app.netlify.com/sites/docs-getdbt-com/functions/get-discourse-comments?scope=deploypreview:2577) for reference.

## ⚠️ Additional Info
This PR adds two new environment variables. 
- `DISCOURSE_USER_SYSTEM` - This allows us to use the Discourse system account to create new topics via the API.
- `DISCOURSE_DEVBLOG_API_KEY` - This new API key has the correct permissions to allow reading, writing, and searching through the API. The current `API_KEY` env variable does not have create permissions.
- Please note that we are currently using the `Lounge` category in Discourse to test this functionality on. This will need to be updated prior to merging as the public can not access this category,
- Introduces [sanitize-html](https://www.npmjs.com/package/sanitize-html) library to prevent XSS Snyk errors when rendering data from API.

## 🚨 TODO Before Merging
- [ ] ~~Set category ID to post new topics to~~

john-rock committed:
wip

john-rock committed:
first push

john-rock committed:
Merge branch 'current' of https://github.com/dbt-labs/docs.getdbt.com into blog-discourse-comments

john-rock committed:
fix window err

john-rock committed:
attempt to fix qs not being found on deploy prev

john-rock committed:
add query params to production path

john-rock committed:
some refactoring

john-rock committed:
add reusable functions

john-rock committed:
add button to start discussion and some more cleanup

john-rock committed:
fix xss err

john-rock committed:
xss fix

john-rock committed:
fix topic id from being undefined

john-rock committed:
add null as default topicId value

john-rock committed:
add some default state

john-rock committed:
add null check

john-rock committed:
begin writing styles

john-rock committed:
add styles

john-rock committed:
cmall changes

john-rock committed:
sanitize discourse url

john-rock committed:
try sanitize package tahat doesnt break next

john-rock committed:
trigger rebuild

john-rock committed:
fix undefined topicId in function

john-rock committed:
add slug as useEffect dep

john-rock committed:
remove console logs

john-rock committed:
sanitize comment html

john-rock committed:
add test

john-rock committed:
Merge branch 'current' into blog-discourse-comments

john-rock committed:
allow emojis in comment body

JKarlavige commented:
Before looking into the code, have a question on this:
> Please note that we are currently using the Lounge category in Discourse to test this functionality on. This will need to be updated prior to merging as the public can not access this category,

and the to-do:
> Set category ID to post new topics to

When ready for take-live, will it dynamically add the discussion to the category of the blog post, or how will it know what category to add it to?

john-rock commented:
> 

The category ID we are posting new topics to in Discourse is set [here](https://github.com/dbt-labs/docs.getdbt.com/pull/2577/files#diff-103fe2b48e2acdd08e0b474acceb780c8735bb3b56eb17df4e306e48f15a2fe7R91-R96). Note the `category: 2`.

We get the category id from the url of the Discourse category (https://discourse.getdbt.com/c/lounge/2).

Once we are ready to go live we will be switching that ID to `21` to post in the In-Depth discussion (https://discourse.getdbt.com/c/discussions/21). @joellabes you can confirm that is the correct category these topics will be posted in once we are ready to go live.

JKarlavige commented on website/functions/get-discourse-comments.js:
```diff
@@ -0,0 +1,154 @@
+    console.log('topicExists', topicExists)
+
+    // If it does not exist in Discourse, create a new topic
+    if(!topicExists) {
+        await createDiscourseTopic(postTitle, postSlug)
+
+        topics = await searchDiscourseTopics(postTitleEncoded)
+
+        if(topics && topics?.length > 0 ) {
+            topicId = await getTopicId(topics, postTitle)
+        }
+
+        if(topicId) {
+            comments = await getDiscourseTopicbyID(topicId)
+        }
+     
+        // Remove the the first post of the comments array since it is not a comment
+        if (comments.length) {
+            comments.shift()
+            comments = { topicId, comments }
+        }
+
+        return await returnResponse(200, comments)
+    } else {
```
Just a❔for when `!topicExists`. Does this flow create a new topic, then search the api to get that newly created `topicId`, which is used to get the comments from there?

If so, i'm wondering if the discourse create topic endpoint happens to return the new topic id when created? Would it be possible to pull the `topicId` directly from the `await createDiscourseTopic(postTitle, postSlug)` response, and pass it into `getDiscourseTopicbyID(topicId)` from there? That might allow lines 48-52 to be eliminated, but not sure if possible, or if those lines are required for anything else.

john-rock replied:
Yes and great call returning the topicId on the `createDiscourseTopic` instead of having another function ran.

Just pushed a change to return topicId from `createDiscourseTopic`, removed the not needed search function, and also updated the Discourse category ID to be a constant.

JKarlavige commented on website/src/components/discourseBlogComments/index.test.js:
```diff
@@ -0,0 +1,154 @@
+  },
+};
+
+describe("Test DiscourseBlogComments component", () => {
```
Nice work with the tests!

JKarlavige reviewed (dismissed):
Nice work with this, started a new discussion on [this blog post](https://deploy-preview-2577--docs-getdbt-com.netlify.app/blog/sam-foundry-experience) and the flow looks to have worked as expected. It created a new Discourse topic, and the page pulls in new comments when added in Discourse. Ran the tests and verified all passed.

Just left one comment with a question, but that is a nonblocker. lgtm!

JKarlavige commented:
> The category ID we are posting new topics to in Discourse is set [here](https://github.com/dbt-labs/docs.getdbt.com/pull/2577/files#diff-103fe2b48e2acdd08e0b474acceb780c8735bb3b56eb17df4e306e48f15a2fe7R91-R96). Note the `category: 2`.
> 
> We get the category ID from the url of the Discourse category (https://discourse.getdbt.com/c/lounge/2).
> 
> Once we are ready to go live we will be switching that ID to `21` to post in the In-Depth discussion (https://discourse.getdbt.com/c/discussions/21). @joellabes you can confirm that is the correct category these topics will be posted in once we are ready to go live.

Awesome, that makes sense. I like the message you create the new topics with which link back to the blog post, wasn't sure where that was coming from initially when looking at the discourse post.

john-rock committed:
return topic_id from createTopic function for cleaner code

john-rock committed:
make dicourse topic id a variable

JKarlavige reviewed (dismissed):
lgtm! verified a post with an existing discourse topic pulls in the comments successfully, and was able to create a new topic for a post without existing comments. all looks good monitoring the logs while going through the steps, nice work!

joellabes commented on website/src/components/discourseBlogComments/index.js:
```diff
@@ -0,0 +1,96 @@
+                  <div dangerouslySetInnerHTML={{__html: sanitizeHtml(comment.cooked, {allowedTags: sanitizeHtml.defaults.allowedTags.concat([ 'img' ])})}} />
+                </li>
+              ))}
+               <a href={sanitizeHtml(`${DISCOURSE_TOPIC_ENDPOINT}${topicId}`)} target="_blank" rel="noopener noreferrer" title='Continnue discussion' className={`button button--primary ${styles.discourseCta}`}>Continue discussion</a>
```
```suggestion
               <a href={sanitizeHtml(`${DISCOURSE_TOPIC_ENDPOINT}${topicId}`)} target="_blank" rel="noopener noreferrer" title='Continue discussion' className={`button button--primary ${styles.discourseCta}`}>Continue discussion</a>
```

joellabes reviewed (dismissed):
The change request is for the typo, the rest of this is non-blocking comments depending on your capacity to make more changes!

I don't feel _great_ about using the title of the post as the match method. Discourse does allow you to [define an embed_url when you create a post](https://docs.discourse.org/#tag/Posts/operation/createTopicPostPM), and then [query for it](https://meta.discourse.org/t/embed-discourse-comments-on-another-website-via-javascript/31963#programmatically-querying-the-embed-details-8), which I think would be more robust. Disclaimer: I haven't tried this, it might run into a problem since we're implementing our own take. 

Are you creating the topic on pageload, or when the user interacts with the comments section? Wondering if we might wind up with a bunch of empty topics on the Discourse side 🤔. Maybe we could unlist the topics when first created, and only change their visibility back to public once a comment is added?

Have you checked what happens if someone writes an enormous post or drops a massive code block? Or if there are 100 comments? Will we display them all? 

lmk what you think! Looks good otherwise 👍

john-rock commented:
Hey @joellabes this is great feedback thank you! I think all of these are worth exploring and left a few comments on each  below.

> I don't feel _great_ about using the title of the post as the match method. Discourse does allow you to [define an embed_url when you create a post](https://docs.discourse.org/#tag/Posts/operation/createTopicPostPM), and then [query for it](https://meta.discourse.org/t/embed-discourse-comments-on-another-website-via-javascript/31963#programmatically-querying-the-embed-details-8), which I think would be more robust. Disclaimer: I haven't tried this, it might run into a problem since we're implementing our own take.

This is interesting and agree may be more robust instead of querying for the title. I'm going to look into seeing if we can work with the embed_url instead of the title.

> Are you creating the topic on pageload, or when the user interacts with the comments section? Wondering if we might wind up with a bunch of empty topics on the Discourse side 🤔. Maybe we could unlist the topics when first created, and only change their visibility back to public once a comment is added?

In its current configuration the Discourse topic is created when the dev blog post is loaded. I think unlisting at first and listing once a comment is added would be better UX within Discourse itself. I'll work on that as well.


> Have you checked what happens if someone writes an enormous post or drops a massive code block? Or if there are 100 comments? Will we display them all?

Great point - I don't have any pagination setup for the number of comments so all comments would be rendered. I think it's worth setting up some type of pagination or maybe even a "load more" button whenever the comments are over a certain number. I'll work on that!

john-rock committed:
fix typo

john-rock committed:
add exteral_url property

john-rock committed:
refactor querying discourse by external_id

john-rock committed:
add styles to set height of code block

john-rock committed:
add suffix to console

john-rock committed:
merge main

john-rock commented:
Hey everyone I've had some time to get back to this over the past day and wanted to provide a quick update of where I'm at.

Regarding querying by external_id instead of title:
- I found out that we can both set and query topics by `external_id`
- However, the once the `external_id` is set in Discourse it remains persistent even after a topic is deleted. At first I tried using our slugs for the `external_id`. This caused errors when trying to query the posts we originally tested when using the iFrame implementation because technically they already exist in Discourse.
- In order to work around this I implemented a `DISCOURSE_EXTERNAL_ID_SUFFIX` that I am appending to the end of our slug when setting and querying an external_id. See [here](https://github.com/dbt-labs/docs.getdbt.com/pull/2577/files#diff-103fe2b48e2acdd08e0b474acceb780c8735bb3b56eb17df4e306e48f15a2fe7R58-R68) for setting and [here](https://github.com/dbt-labs/docs.getdbt.com/pull/2577/files#diff-103fe2b48e2acdd08e0b474acceb780c8735bb3b56eb17df4e306e48f15a2fe7R94-R102) for query. I have it set as a testing value (-testing2) for now that I will update to a more appropriate value before going live.

As an added bonus, now that we are setting the `embed_url` the topic that gets created in Discourse will now show a "Show Full Post" button when clicked will show the entire dev blog post contents within Discourse! [See here](https://discourse.getdbt.com/t/how-to-move-data-from-spreadsheets-into-your-data-warehouse/6487) for an example.

I'm still working through the other two callouts and will request review again once I have them in place.

john-rock committed:
add styles for discourse elements

john-rock committed:
cleanup styles

john-rock committed:
add pagiantion when there are more than 6 comments

john-rock committed:
minor refactor

john-rock committed:
refactor to set variable value by default

john-rock committed:
remove trim

john-rock committed:
remove consoles

john-rock committed:
make external id less than 50 characters per discourse requirement

john-rock committed:
cleanup

john-rock committed:
make topic hidden by default and visible when a comment is added

john-rock commented:
> Have you checked what happens if someone writes an enormous post or drops a massive code block? Or if there are 100 comments? Will we display them all?

- Just added support for only showing 6 comments by default and if there are more than 6 a "View more comments" button will render. Each time the button is clicked it will show 6 more comments until all of the comments are visible. 6 is just an arbitrary number I used let me know if we would like to have that adjusted. [Example here](https://deploy-preview-2577--docs-getdbt-com.netlify.app/blog/intelligent-slim-ci).
- Code blocks are now also capped at 400px and will show scroll bars when inner content is > 400px.

> Are you creating the topic on pageload, or when the user interacts with the comments section? Wondering if we might wind up with a bunch of empty topics on the Discourse side 🤔. Maybe we could unlist the topics when first created, and only change their visibility back to public once a comment is added?

This feature is now implemented with the following flow:
- When a dev blog page is loaded, a companion topic will be created in Discourse with a visibility of hidden.
- When a comment is added, the topic will become visible.

At this point I'm opening for another round of review. Let me know if y'all have questions!

john-rock committed:
small cleanup

joellabes commented on website/functions/get-discourse-comments.js:
```diff
@@ -12,117 +13,107 @@ let headers = {
+        }
+
+        // Remove any posts that are empty. This is usually an action set in Discourse (e.g. unlisting/listing a topic).
+        post_stream.posts = post_stream.posts.filter(post => post.cooked)
```
Looks like this can be `post_type == 1`: https://meta.discourse.org/t/about-post-type-in-api-responses/180123

john-rock replied:
Good call! I just added that in.

john-rock replied:
Good call! I just added that in.

JKarlavige reviewed (dismissed):
This lgtm! Able to start a new discussion on a [blog post](https://deploy-preview-2577--docs-getdbt-com.netlify.app/blog/marie-kondo-query-migration). Verified comment showing on blog page after posting in Discourse. Tested the "view more comments" and all seems to work well.

:shipit:

joellabes commented on website/functions/get-discourse-comments.js:
```diff
@@ -3,6 +3,7 @@ const axios = require('axios')
 const { DISCOURSE_DEVBLOG_API_KEY , DISCOURSE_USER_SYSTEM } = process.env
 const DEVBLOG_URL = 'https://docs.getdbt.com/blog/'
 const DISCOURSE_TOPIC_ID = 2
+const DISCOURSE_EXTERNAL_ID_SUFFIX = '-testing3'
```
Are we thinking that this suffix would be blank in prod? If you let me know which slugs are already burned from testing I can try and hard delete them or get the support team to hard delete them

john-rock replied:
Ideally yes, in production we wouldn't have to use a suffix. As it sits now we do have to use the suffix to differentiate between any stale external_ids that are still within Discourse.

 Let me see if I can pull a list of the external_ids that are currently set in Discourse.

john-rock committed:
add post_type for post filtering

joellabes reviewed (dismissed):
Approved (pending the details of the already-burned discourse topics)

Alternatively could we undelete them and delete their test posts? I can move posts from one category to another @john-rock

joellabes commented:
@john-rock Thanks for sticking with it! 

For those two topics you listed, I have changed their names in discourse so that you can use them. If it happens in the future then yes appending ` | dbt Developer Blog` is a good workaround. 

Surprised that you're getting clashes with "how we structure our projects" though since that's a Guide not a blog post. I'm not necessarily against it, just didn't expect it

john-rock commented:
> For those two topics you listed, I have changed their names in discourse so that you can use them. If it happens in the future then yes appending  | dbt Developer Blog is a good workaround.
> 
> Surprised that you're getting clashes with "how we structure our projects" though since that's a Guide not a blog post. I'm not necessarily against it, just didn't expect it

Now that the title error is cleaned up, I'm getting a `errors: [ 'Embed url has already been taken' ]` on the same posts linked above even with a unique `external_id`. I'm guessing those were in the batch we tested with the iframe comment implementation. 

This is going to be tricky because ideally I think we want to keep the embed_url that way the full content of the dev blog post is brought into Discourse. It looks like [deleting the associated topic](https://meta.discourse.org/t/error-message-embed-url-already-being-used-even-after-unlinking-post-from-discourse/153082/2) in Discourse should remove the error but we obviously don't want to do that unless there is a way to dissociate a topic from an embed_url without deleting the topic.

I'm thinking we should only run into this error on older posts. And if that is the case I'm wondering if we can return a message on the frontend with something along the lines of "Topic archive, comments disabled" although it looks like you may still be able to comment on archived posts within Discourse 🤔

joellabes commented:
I have poked around in the Data Explorer (their Query Interface over the tables) and have a sample query here: https://discourse.getdbt.com/admin/plugins/explorer?id=2

There are ~120 topics, but they don't contain any comments from anyone other than me, you and Jason K so I've emailed the Discourse team and asked them to delete them all so we can start from scratch.

While I was there, I noticed that the hash/anchor part of the page was also triggering duplicate posts to be created, e.g. [3802](https://discourse.getdbt.com/t/your-essential-dbt-project-checklist-modeling-dbt-community-discourse/3802) and [3803](https://discourse.getdbt.com/t/your-essential-dbt-project-checklist-modeling-dbt-community-discourse/3803) are the same topic. 

I also found [3796](https://discourse.getdbt.com/t/how-to-build-a-mature-dbt-project-from-scratch-dbt-developer-blog/3796) which has a hubspot tracking ID in it (https://docs.getdbt.com/blog/how-to-build-a-mature-dbt-project-from-scratch?_hsenc=p2anqtz-_k9b8rrevdyrmak6snv-u3ijrujg3xykdzycq5hmiv9wg_yblw4ffzanah590y8oh6uku_) which is weird! It was created yesterday which makes me think that we must already have something live?

john-rock commented:
Ope - actually looks like [3796](https://discourse.getdbt.com/t/how-to-build-a-mature-dbt-project-from-scratch-dbt-developer-blog/3796) was created Feb 5, 2022.

![image](https://user-images.githubusercontent.com/46692803/217278545-1526d8a2-0d02-414b-b1c5-27bc78ed4e9b.png)

john-rock committed:
add function to clean hash and parms from url to prevent dupe topics in discourse

john-rock committed:
testing using context to build payload for api to prevent dupelicate errors while testing

john-rock committed:
add context inside function

john-rock committed:
testing

john-rock committed:
add deploy context vars

john-rock committed:
fix env in toml

john-rock committed:
add empty var

john-rock committed:
whoops

john-rock committed:
refactor testing logic

john-rock committed:
clean up setting variables

john-rock committed:
cleanup

john-rock committed:
empty commit

john-rock committed:
remove extra files

john-rock committed:
merge current

john-rock committed:
remove non needed items in netlify.toml

john-rock committed:
add back function to create topic

john-rock committed:
small refactor for blogUrl

john-rock committed:
add check for production so we dont need to switch categorys

john-rock committed:
small cleanup

john-rock commented:
Alrighty so two things I added with the last few commits:

1. Added function to remove any query or hash params within the URL to prevent duplicate topics from being created.
2. Added a check to determine which environment we are on and adding a unique `external_id` if we are either on a local dev environment or a deploy preview. This will allow us to test locally and within deploy previews without creating a conflict in the production environment.
```
    const env =
      blogUrl === DEVBLOG_PROD_URL
        ? ""
        : blogUrl.includes("localhost")
        ? DEV_ENV
        : PREVIEW_ENV;
    postTitle = `${env}${event.queryStringParameters.title}`;
    postSlug = event.queryStringParameters.slug;
    cleanSlug = cleanUrl(postSlug);
    externalId = truncateString(`${env}${cleanSlug}`);
```
For example, if we access [this post](https://deploy-preview-2577--docs-getdbt-com.netlify.app/blog/how-to-create-near-real-time-models-with-just-dbt-sql#write-idempotent-models) in a deploy preview, the `external_id` will be `deploy-preview-how-to-create-near-real-time-models` and in production it would be `how-to-create-near-real-time-models`.

3. Added a check to post all topics in the Lounge category within Discourse if created in a local environment or deploy preview and posted in In-Depth Discussions if created in a production environment. The topic will still show unlisted until the first comment is added.
```
  if (blogUrl === DEVBLOG_PROD_URL) {
    DISCOURSE_TOPIC_ID = 21
  } else {
    DISCOURSE_TOPIC_ID = 2
  }
```

@joellabes I believe with this in place we shouldn't let run into creating conflicts while developing or looking at a deploy preview. Take a look and let me know what you think. If it looks good, and once the topics are hard deleted in discourse, we should be good to go.

joellabes commented:
> Ope - actually looks like [3796](https://discourse.getdbt.com/t/how-to-build-a-mature-dbt-project-from-scratch-dbt-developer-blog/3796) was created Feb 5, 2022.

DOH! my bad, sorry 😬 I clearly didn't read to the year part of the stamp 🤦

Having a look at this now, but your changeset sounds great!

joellabes committed:
Merge branch 'current' into blog-discourse-comments

joellabes commented:
The diff is hella messy because there's been some new articles merged in, but this looks good to me! The Discourse team replied to me overnight checking I really really wanted these to be hard-deleted, and I replied saying yes please this morning; I imagine they will action it overnight for me and we should be good to go tomorrow 🥳

joellabes commented:
Posts are deleted!

JKarlavige commented on website/functions/get-discourse-comments.js:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  try {
+    const env =
+      blogUrl === DEVBLOG_PROD_URL
+        ? ""
+        : blogUrl.includes("localhost")
+        ? DEV_ENV
+        : PREVIEW_ENV;
+    postTitle = `${env}${event.queryStringParameters.title}`;
+    postSlug = event.queryStringParameters.slug;
+    cleanSlug = cleanUrl(postSlug);
+    externalId = truncateString(`${env}${cleanSlug}`);
+
+    console.table({
+      blogUrl,
+      postTitle,
+      postSlug,
+      cleanSlug,
+      externalId,
+    });
```
This is nice and concise!

JKarlavige reviewed (approved):
All lgtm! Verified able to add comments to existing blog post within Discourse and new comments appear on the post. Able to create a new Discourse topic for a post with no comments, and new comment appears on post afterwards.

Nice job with this, it looks like a lot of work went into it!

john-rock committed:
Merge branch 'current' into blog-discourse-comments
